qoqo-macros/src/noise_models.rs

```diff
@@ -0,0 +1,144 @@
+// express or implied. See the License for the specific language governing permissions and
+// limitations under the License.
+
+use quote::quote;
+use syn::{parse_macro_input, ItemImpl};
+
+pub fn noise_model_wrapper_def(
+    _metadata: proc_macro::TokenStream,
+    input: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let parsed_input = parse_macro_input!(input as ItemImpl);
+    let ident = parsed_input.self_ty;
+
+    let items = parsed_input.items;
+    let q = quote! {
+        #[pymethods]
+        impl #ident {
+            #(#items)*
+
+            /// Returns a copy of the device (copy here produces a deepcopy).
+            ///
+            /// Returns:
+            ///     A deep copy of self.
+            ///
+            pub fn __copy__(&self) -> Self {
+                self.clone()
+            }
+
+            /// Creates deep copy of Noise-Model.
+            ///
+            /// Returns:
+            ///     A deep copy of self.
+            ///
+            pub fn __deepcopy__(&self, _memodict: Py<PyAny>) -> Self {
+                self.clone()
+            }
+
+            /// Return the bincode representation of the Noise-Model using the bincode crate.
+            ///
+            /// Returns:
+            ///     ByteArray: The serialized Noise-Model (in bincode form).
+            ///
+            /// Raises:
+            ///     ValueError: Cannot serialize Noise-Model to bytes.
+            ///
+            pub fn to_bincode(&self) -> PyResult<Py<pyo3::types::PyByteArray>> {
+                let noise_model = NoiseModel::from(self.internal.clone());
+                let serialized = bincode::serialize(&noise_model)
+                    .map_err(|_| pyo3::exceptions::PyValueError::new_err("Cannot serialize Noise-Model to bytes"))?;
+                let b: Py<pyo3::types::PyByteArray> = Python::with_gil(|py| -> Py<pyo3::types::PyByteArray> {
+                    pyo3::types::PyByteArray::new(py, &serialized[..]).into()
+                });
+                Ok(b)
+            }
+
+            /// Return the json representation of the Noise-Model.
+            ///
+            /// Returns:
+            ///     str: The serialized form of Noise-Model.
+            ///
+            /// Raises:
+            ///     ValueError: Cannot serialize Noise-Model to json.
+            ///
+            pub fn to_json(&self) -> PyResult<String> {
+                let noise_model = NoiseModel::from(self.internal.clone());
+                let serialized = serde_json::to_string(&noise_model)
+                    .map_err(|_| pyo3::exceptions::PyValueError::new_err("Cannot serialize Noise-Model to json"))?;
+                Ok(serialized)
+            }
+
+
+
+
+
+            fn __repr__(&self) -> String{
+                format!("{:?}", self.internal)
+            }
+
+
+            /// Return the __richcmp__ magic method to perform rich comparison operations on mixed system.
```

Comment:
typo from mixed systems

Reply:
Not tested, because that was actually not meant to be here.